---
title: "filtration_rat_cerebellum"
output: html_document
date: "2023-09-29"
editor_options: 
  chunk_output_type: console
---

Loading Libraries
```{r warning=FALSE, message=FALSE}
suppressPackageStartupMessages({
  library(tidyverse)
  library(GenomicFeatures)
  library(rtracklayer)
  library(dplyr)
  library(Rsamtools)
  library(glmnet)
  library(ggpubr)
  library(visreg)
  library(DESeq2)
  library(apeglm)
  library(slider)
})
```

loading low and high rin data. 
```{r warning=FALSE, message=FALSE}
lrin_rep1 <- set_1
lrin_rep2 <- set_1
hrin_rep1 <- set_1
hrin_rep2 <- set_1
```

Indels and 3' end filtering

low rin- replicate 1
```{r warning=FALSE, message=FALSE}
# Del filtering
filter1l <- lrin_rep1 %>% 
  mutate(filter_1=ifelse(MapLen > 200 & MaxDelLen > 160 , "discard", "not discard"))
keep1l <- filter1l %>% 
  filter(filter_1=="not discard")

# Insert filtering
filter2l <- keep1l %>% 
  mutate(filter_2=ifelse(MapLen > 200 & MaxInsertLen > 80 , "discard", "not discard"))
keep2l <- filter2l %>% 
  filter(filter_2=="not discard")

# Selecting required columns
test1l <- keep2l %>% 
  dplyr::select(Read, Transcript, StartCoord, MapLen, DelLen, InsertLen, SoftClipLenSum, NReads, end_without_sc, ReadLen, MaxDelLen, MaxInsertLen, MaxSoftClipLen) %>% 
  dplyr::group_by(Transcript)

# 3' end filtering
test_l_order <- test1l[order(test1l$end_without_sc, decreasing = TRUE),]

test2 <- test_l_order %>% 
  group_by(Transcript, end_without_sc) %>% 
  mutate(number_of_reads_for_each_len=n()) %>% 
  ungroup()
  
test2l <- test2 %>% 
  dplyr::group_by(Transcript) %>% 
  mutate(x=cumsum(number_of_reads_for_each_len))

p <- function(x,n){
  len.diff <- n - length(x)
  c(rep(NA, len.diff), x)
}

test2l1 <- test2l %>% 
  dplyr::group_by(Transcript) %>% 
  mutate(dif=p(diff(x, lag=20), length(x)))

test2l1$dif[is.na(test2l1$dif)] <- 0

test2l2 <- test2l1 %>% 
  dplyr::group_by(Transcript) %>% 
  mutate(saturation=end_without_sc[which.max(dif)])

#Subtracting saturation level with 3' end
test2l2$E <- test2l2$saturation - test2l2$end_without_sc
test2l2$abs_E <- abs(test2l2$E)

# Keeping reads that are absE < 50
filter3l <- test2l2 %>% 
  filter(abs_E < 50)

# Calculating new 3'end 
filter3l$fragment_len <- filter3l$saturation - filter3l$StartCoord
```

Selecting required columns
```{r warning=FALSE, message=FALSE}
colnames(filter3l)[19] <- "new_3_end"
colnames(filter3l)[20] <- "new_ReadLen"

lrin_all_filtering_rep1 <- filter3l %>% 
  dplyr::select(Read, Transcript, StartCoord, MapLen, DelLen, InsertLen, SoftClipLenSum, NReads, end_without_sc, ReadLen, MaxDelLen, MaxInsertLen, MaxSoftClipLen, saturation, new_3_end, new_ReadLen)
```

low rin- replicate 2
```{r warning=FALSE, message=FALSE}
# Del filtering
filter1l <- lrin_rep2 %>% 
  mutate(filter_1=ifelse(MapLen > 200 & MaxDelLen > 160 , "discard", "not discard"))
keep1l <- filter1l %>% 
  filter(filter_1=="not discard")

# Insert filtering
filter2l <- keep1l %>% 
  mutate(filter_2=ifelse(MapLen > 200 & MaxInsertLen > 80 , "discard", "not discard"))
keep2l <- filter2l %>% 
  filter(filter_2=="not discard")

# Selecting required columns
test1l <- keep2l %>% 
  dplyr::select(Read, Transcript, StartCoord, MapLen, DelLen, InsertLen, SoftClipLenSum, NReads, end_without_sc, ReadLen, MaxDelLen, MaxInsertLen, MaxSoftClipLen) %>% 
  dplyr::group_by(Transcript)

# 3' end filtering
test_l_order <- test1l[order(test1l$end_without_sc, decreasing = TRUE),]

test2 <- test_l_order %>% 
  group_by(Transcript, end_without_sc) %>% 
  mutate(number_of_reads_for_each_len=n()) %>% 
  ungroup()
  
test2l <- test2 %>% 
  dplyr::group_by(Transcript) %>% 
  mutate(x=cumsum(number_of_reads_for_each_len))

p <- function(x,n){
  len.diff <- n - length(x)
  c(rep(NA, len.diff), x)
}

test2l1 <- test2l %>% 
  dplyr::group_by(Transcript) %>% 
  mutate(dif=p(diff(x, lag=20), length(x)))

test2l1$dif[is.na(test2l1$dif)] <- 0

test2l2 <- test2l1 %>% 
  dplyr::group_by(Transcript) %>% 
  mutate(saturation=end_without_sc[which.max(dif)])

#Subtracting saturation level with 3' end
test2l2$E <- test2l2$saturation - test2l2$end_without_sc
test2l2$abs_E <- abs(test2l2$E)

# Keeping reads that are absE < 50
filter3l <- test2l2 %>% 
  filter(abs_E < 50)

# Calculating new 3'end 
filter3l$fragment_len <- filter3l$saturation - filter3l$StartCoord
```

Selecting required columns
```{r warning=FALSE, message=FALSE}
colnames(filter3l)[19] <- "new_3_end"
colnames(filter3l)[20] <- "new_ReadLen"

lrin_all_filtering_rep2 <- filter3l %>% 
  dplyr::select(Read, Transcript, StartCoord, MapLen, DelLen, InsertLen, SoftClipLenSum, NReads, end_without_sc, ReadLen, MaxDelLen, MaxInsertLen, MaxSoftClipLen, saturation, new_3_end, new_ReadLen)
```

high rin- replicate 1
```{r warning=FALSE, message=FALSE}
# Del filtering
filter1h <- hrin_rep1 %>% 
  mutate(filter_1=ifelse(MapLen > 200 & MaxDelLen > 160 , "discard", "not discard"))
keep1h <- filter1h %>% 
  filter(filter_1=="not discard")

# Insert filtering
filter2h <- keep1h %>% 
  mutate(filter_2=ifelse(MapLen > 200 & MaxInsertLen > 80 , "discard", "not discard"))
keep2h <- filter2h %>% 
  filter(filter_2=="not discard")

# Selecting required columns
test1h <- keep2h %>% 
  dplyr::select(Read, Transcript, StartCoord, MapLen, DelLen, InsertLen, SoftClipLenSum, NReads, end_without_sc, ReadLen, MaxDelLen, MaxInsertLen, MaxSoftClipLen)

# 3' end filtering
test_h_order <- test1h[order(test1h$end_without_sc, decreasing = TRUE),]

test2h <- test_h_order %>% 
  group_by(Transcript, end_without_sc) %>% 
  mutate(number_of_reads_for_each_len=n()) %>% 
  ungroup()
  
test2h_set1 <- test2h %>% 
  dplyr::group_by(Transcript) %>% 
  mutate(x=cumsum(number_of_reads_for_each_len))

p <- function(x,n){
  len.diff <- n - length(x)
  c(rep(NA, len.diff), x)
}

test2h_set2 <- test2h_set1 %>% 
  dplyr::group_by(Transcript) %>% 
  mutate(dif=p(diff(x, lag=20), length(x)))

test2h_set2$dif[is.na(test2h_set2$dif)] <- 0

test2h_set3 <- test2h_set2 %>% 
  dplyr::group_by(Transcript) %>% 
  mutate(saturation=end_without_sc[which.max(dif)])

#Subtracting saturation level with 3' end
test2h_set3$E <- test2h_set3$saturation - test2h_set3$end_without_sc
test2h_set3$abs_E <- abs(test2h_set3$E)

# Keeping reads that are absE < 50
filter3h <- test2h_set3 %>% 
  filter(abs_E < 50)

# Calculating new 3'end 
filter3h$fragment_len <- filter3h$saturation - filter3h$StartCoord
```

Selecting required columns
```{r warning=FALSE, message=FALSE}
colnames(filter3h)[19] <- "new_3_end"
colnames(filter3h)[20] <- "new_ReadLen"

hrin_all_filtering_rep1 <- filter3h %>% 
  dplyr::select(Read, Transcript, StartCoord, MapLen, DelLen, InsertLen, SoftClipLenSum, NReads, end_without_sc, ReadLen, MaxDelLen, MaxInsertLen, MaxSoftClipLen, saturation, new_3_end, new_ReadLen)
```

high rin- replicate 2
```{r warning=FALSE, message=FALSE}
# Del filtering
filter1h <- hrin_rep2 %>% 
  mutate(filter_1=ifelse(MapLen > 200 & MaxDelLen > 160 , "discard", "not discard"))
keep1h <- filter1h %>% 
  filter(filter_1=="not discard")

# Insert filtering
filter2h <- keep1h %>% 
  mutate(filter_2=ifelse(MapLen > 200 & MaxInsertLen > 80 , "discard", "not discard"))
keep2h <- filter2h %>% 
  filter(filter_2=="not discard")

# Selecting required columns
test1h <- keep2h %>% 
  dplyr::select(Read, Transcript, StartCoord, MapLen, DelLen, InsertLen, SoftClipLenSum, NReads, end_without_sc, ReadLen, MaxDelLen, MaxInsertLen, MaxSoftClipLen)

# 3' end filtering
test_h_order <- test1h[order(test1h$end_without_sc, decreasing = TRUE),]

test2h <- test_h_order %>% 
  group_by(Transcript, end_without_sc) %>% 
  mutate(number_of_reads_for_each_len=n()) %>% 
  ungroup()
  
test2h_set1 <- test2h %>% 
  dplyr::group_by(Transcript) %>% 
  mutate(x=cumsum(number_of_reads_for_each_len))

p <- function(x,n){
  len.diff <- n - length(x)
  c(rep(NA, len.diff), x)
}

test2h_set2 <- test2h_set1 %>% 
  dplyr::group_by(Transcript) %>% 
  mutate(dif=p(diff(x, lag=20), length(x)))

test2h_set2$dif[is.na(test2h_set2$dif)] <- 0

test2h_set3 <- test2h_set2 %>% 
  dplyr::group_by(Transcript) %>% 
  mutate(saturation=end_without_sc[which.max(dif)])

#Subtracting saturation level with 3' end
test2h_set3$E <- test2h_set3$saturation - test2h_set3$end_without_sc
test2h_set3$abs_E <- abs(test2h_set3$E)

# Keeping reads that are absE < 50
filter3h <- test2h_set3 %>% 
  filter(abs_E < 50)

# Calculating new 3'end 
filter3h$fragment_len <- filter3h$saturation - filter3h$StartCoord
```

Selecting required columns
```{r warning=FALSE, message=FALSE}
colnames(filter3h)[19] <- "new_3_end"
colnames(filter3h)[20] <- "new_ReadLen"

hrin_all_filtering_rep2 <- filter3h %>% 
  dplyr::select(Read, Transcript, StartCoord, MapLen, DelLen, InsertLen, SoftClipLenSum, NReads, end_without_sc, ReadLen, MaxDelLen, MaxInsertLen, MaxSoftClipLen, saturation, new_3_end, new_ReadLen)
```