---
title: "separate_lens_plots"
output: html_document
date: "2023-04-04"
editor_options: 
  chunk_output_type: console
---

This script is for the: 
1. separate lens showing where the 95% of the reads are for all three subset of the datasets. 
2. The plots also have the correlation, p-value and the slope of the fitted linear model. 
3. This also includes the code that can tell the number of reads above and below the linear model line.
4. This also includes the comparison of reads with Pysam. 


Loading Libraries
```{r warning=FALSE, message=FALSE}
suppressPackageStartupMessages({
  library(tidyverse)
  library(GenomicFeatures)
  library(rtracklayer)
  library(dplyr)
  library(Rsamtools)
  library(glmnet)
  library(ggpubr)
  library(visreg)
})
```

Loading the subset data- Bhavika's Laptop
```{r warning=FALSE, message=FALSE}
undegraded <- read.csv("d:/Users/Sujata Kumar/Desktop/Nanograd/Data/undegraded_subset.csv")
degraded_heavy <- read.csv("d:/Users/Sujata Kumar/Desktop/Nanograd/Data/subset_degraded.csv")
```

Loading the subset data- JCSMR Machine
```{r warning=FALSE, message=FALSE}
undegraded <- read.csv("/home/bhavika/Desktop/Nanograd/Data/csv_files/subset_un.csv")
degraded_heavy <- read.csv("/home/bhavika/Desktop/Nanograd/Data/csv_files/subset_heavy.csv")
mildly_degraded <- read.csv("/home/bhavika/Desktop/Nanograd/Data/csv_files/subset_md.csv")
```

made the subset of the bam file using samtools on gadi
```{r warning=FALSE, message=FALSE}
b <- a %>% 
  filter(Read=="e48d58ca-c16f-4d3e-99a0-65fe7e8517c9")
```

Loading Pysam output to cross check and compare with the end_coord function
```{r warning=FALSE, message=FALSE}
pysam <- read.csv("/home/bhavika/Desktop/Nanograd/Data/pysam_output.csv")

pysam <- read.csv("d:/Users/Sujata Kumar/Desktop/Nanograd/Data/pysam_output.csv")

c <- pysam %>% 
  filter(ReadName == "e48d58ca-c16f-4d3e-99a0-65fe7e8517c9")
```

##Degraded sample

1. Map Len vs Del Len
Plot with both the curves- lm and loess
```{r warning=FALSE, message=FALSE}
ggplot(subset_final, aes(x=MapLen, y=DelLen)) + geom_point() + labs(x="Map Len", y="Del Len") + theme_bw() + geom_smooth(method = "lm") + geom_smooth(method = "loess", colour="green") + ylim(0,400)
```

Specifying 95% region
```{r warning=FALSE, message=FALSE}
ggplot(subset_final, aes(x=MapLen, y=DelLen)) + geom_point() + labs(x="Map Len", y="Del Len") + theme_bw() + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.95, colour="red") + geom_smooth(method = "lm") + ylim(0,400) + geom_hline(yintercept = 80, colour="green") + geom_vline(xintercept = 1050, colour="green")
```

Adding correlation, slope, p-values and intercept to the plot
Plot with ggpubr library
```{r warning=FALSE, message=FALSE}
model_1 <- lm(DelLen ~ MapLen, data = degraded_heavy)

ggscatter(degraded_heavy, x="MapLen", y="DelLen", add = "reg.line", conf.int = T, cor.coef = T, cor.method = "pearson", colour="blue") + ylim(0,400) + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.95, colour="red") +  geom_hline(yintercept = 80, colour = "green") + geom_vline(xintercept = 1050, colour="green") + labs(title = paste("Intercept=", signif(model_1$coefficients[[1]]), "Slope=", signif(model_1$coefficients[[2]])))
```


Showing 90% region
```{r warning=FALSE, message=FALSE}
ggplot(degraded_heavy, aes(x=MapLen, y=DelLen)) + geom_point() + labs(x="Map Len", y="Del Len") + theme_bw() + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.90, colour="red") + geom_smooth(method = "lm") + ylim(0,400) + geom_hline(yintercept = 75, colour="green") + geom_vline(xintercept = 980, colour="green")
```

How many genes are above or below the lm line
```{r warning=FALSE, message=FALSE}
model_1 <- lm(DelLen ~ MapLen, data = subset_final)
b <- resid(model_1)
sum(b<0)
sum(b>0)
```

2. Map Len vs Insert Len
```{r warning=FALSE, message=FALSE}
ggplot(subset_final, aes(x=MapLen, y=InsertLen)) + geom_point() + labs(x="Map Len", y="Insert Len") + theme_bw() + geom_smooth(method = "lm") + geom_smooth(method = "loess", colour="green") + xlim(0,2500) + ylim(0,100)
```

Number of genes below or above the lm line
```{r warning=FALSE, message=FALSE}
model_2 <- lm(InsertLen ~ MapLen, data = subset_final)
b <- resid(model_2)
sum(b<0)
sum(b>0)
```

Showing the region with the cut offs
```{r warning=FALSE, message=FALSE}
test <- degraded %>% 
  filter(InsertLen < 200)

ggplot(subset_final, aes(x=MapLen, y=InsertLen)) + geom_point() + labs(x="Map Len", y="Insert Len") + theme_bw() + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.95, colour="red") + geom_smooth(method = "lm") + xlim(0,2500) + ylim(0,100) + geom_hline(yintercept = 33, colour="green") + geom_vline(xintercept = 1070, colour="green")
```

Adding correlation, slope, p-values and intercept to the plot
Plot with ggpubr library
```{r warning=FALSE, message=FALSE}
model_2 <- lm(InsertLen ~ MapLen, data = degraded_heavy)

ggscatter(degraded_heavy, x="MapLen", y="InsertLen", add = "reg.line", conf.int = T, cor.coef = T, cor.method = "pearson", colour="blue") + ylim(0,100) + xlim(0,2500) + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.95, colour="red") +  geom_hline(yintercept = 33, colour = "green") + geom_vline(xintercept = 1070, colour="green") + labs(title = paste("Intercept=", signif(model_2$coefficients[[1]]),"Slope=", signif(model_2$coefficients[[2]])))
```

Map len vs soft clip len
```{r warning=FALSE, message=FALSE}
ggplot(subset_final, aes(x=MapLen, y=SoftClipLen)) + geom_point() + labs(x="Map Len", y="Soft Clip Len") + theme_bw() + xlim(0,3500) + ylim(0,1000) + geom_smooth(method = "lm") + geom_smooth(method = "loess", colour="green")

ggplot(subset_final, aes(x=MapLen, y=SoftClipLen)) + geom_point() + labs(x="Map Len", y="Soft Clip Len") + theme_bw() + xlim(0,3500) + ylim(0,1000) + geom_hline(yintercept = 150, colour="red") + geom_vline(xintercept = 1000, colour="blue") + geom_vline(xintercept = 2000, colour="blue")

c <- degraded_heavy %>% 
  filter(MapLen > 2000 & SoftClipLen <130)
```

Number of genes below or above the lm line
```{r warning=FALSE, message=FALSE}
model_3 <- lm(SoftClipLen ~ MapLen, data = subset_final)
b <- resid(model_3)
sum(b<0)
sum(b>0)
```

Showing 95% of the regions and finalizing the cutoff
```{r warning=FALSE, message=FALSE}
ggplot(subset_final, aes(x=MapLen, y=SoftClipLen)) + geom_point() + labs(x="Map Len", y="Soft Clip Len") + theme_bw() + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.95, colour="red") + geom_smooth(method = "lm")  + ylim(0,1000) + geom_hline(yintercept = 125, colour="green") + geom_vline(xintercept = 1100, colour="green")
```

```{r warning=FALSE, message=FALSE}
ggplot(degraded_heavy, aes(x=MapLen, y=SoftClipLen)) + geom_point() + labs(x="Map Len", y="Soft Clip Len") + theme_bw() + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.70, colour="red") + geom_smooth(method = "lm") + ylim(0,1000) + geom_hline(yintercept = 95, colour = "green")
```

Adding correlation, slope, p-values and intercept to the plot
Plot with ggpubr library
```{r warning=FALSE, message=FALSE}
model_3 <- lm(SoftClipLen ~ MapLen, data = degraded_heavy)

ggscatter(degraded_heavy, x="MapLen", y="SoftClipLen", add = "reg.line", conf.int = T, cor.coef = T, cor.method = "pearson", colour="blue") + ylim(0,1000) + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.95, colour="red") +  geom_hline(yintercept = 125, colour = "green") + geom_vline(xintercept = 1100, colour="green") + labs(title = paste("Intercept=", signif(model_3$coefficients[[1]]),"Slope=", signif(model_3$coefficients[[2]])))
```


## Read Len
1. Read len vs del len
```{r}
ggplot(degraded_heavy, aes(x=ReadLen, y=DelLen)) + geom_point() + labs(x="Read Len", y="Del Len") + theme_bw() + geom_smooth(method = "lm") + geom_smooth(method = "loess", colour="green") + ylim(0,400)
```

Without loess
```{r warning=FALSE, message=FALSE}
ggplot(degraded_heavy, aes(x=ReadLen, y=DelLen)) + geom_point() + labs(x="Read Len", y="Del Len") + theme_bw() + geom_smooth(method = "lm") + ylim(0,400)
```

Adding correlation, slope, p-values and intercept to the plot
Plot with ggpubr library
```{r warning=FALSE, message=FALSE}
model_1 <- lm(DelLen ~ ReadLen, data = degraded_heavy)

ggscatter(degraded_heavy, x="ReadLen", y="DelLen", add = "reg.line", conf.int = T, cor.coef = T, cor.method = "pearson", colour="blue") + ylim(0,400) + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.95, colour="red") +  geom_hline(yintercept = 80, colour = "green") + geom_vline(xintercept = 1150, colour="green") + labs(title = paste("Intercept=", signif(model_1$coefficients[[1]]), "Slope=", signif(model_1$coefficients[[2]])))
```

Number of genes above or below the lm line
```{r warning=FALSE, message=FALSE}
model_3 <- lm(DelLen ~ ReadLen, data = subset_final)
b <- resid(model_3)
sum(b<0)
sum(b>0)
```

Showing 95% of the reads and finding the cut off
```{r warning=FALSE, message= FALSE}
ggplot(subset_final, aes(x=ReadLen, y=DelLen)) + geom_point() + labs(x="Read Len", y=" Del Len") + theme_bw() + geom_smooth(method = "lm") + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.95, colour="red") + ylim(0,400) + geom_hline(yintercept = 80, colour = "green") + geom_vline(xintercept = 1150, colour="green")
```

Showing 90% of the reads and finding the cutoff
```{r warning=FALSE, message=FALSE}
ggplot(degraded_heavy, aes(x=ReadLen, y=DelLen)) + geom_point() + labs(x="Read Len", y=" Del Len") + theme_bw() + geom_smooth(method = "lm") + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.90, colour="red") + ylim(0,400) + geom_hline(yintercept = 75, colour = "green") + geom_vline(xintercept = 1070, colour="green")
```

2. Read len and Insert len
```{r warning=FALSE, message=FALSE}
ggplot(subset_final, aes(x=ReadLen, y=InsertLen)) + geom_point() + labs(x="Read Len", y="Insert Len") + theme_bw() + geom_smooth(method = "lm") + geom_smooth(method = "loess", colour="green") + xlim(0,2500) + ylim(0,100)
```

Number of genes above and below the lm line
```{r warning=FALSE, message=FALSE}
model_3 <- lm(InsertLen ~ ReadLen, data = subset_final)
b <- resid(model_3)
sum(b<0)
sum(b>0)
```

Showing 95% of the reads and finding the cut off
```{r warning=FALSE, message=FALSE}
ggplot(subset_final, aes(x=ReadLen, y=InsertLen)) + geom_point() + labs(x="Read Len", y="Insert Len") + theme_bw() + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.95, colour="red") + geom_smooth(method = "lm") + ylim(0,100) + xlim(0,2500) + geom_hline(yintercept = 33, colour="green") + geom_vline(xintercept = 1150, colour="green")
```

Adding correlation, slope, p-values and intercept to the plot
Plot with ggpubr library
```{r warning=FALSE, message=FALSE}
model_2 <- lm(InsertLen ~ ReadLen, data = degraded_heavy)

ggscatter(degraded_heavy, x="ReadLen", y="InsertLen", add = "reg.line", conf.int = T, cor.coef = T, cor.method = "pearson", colour="blue") + ylim(0,100) + xlim(0,2500) + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.95, colour="red") +  geom_hline(yintercept = 33, colour = "green") + geom_vline(xintercept = 1150, colour="green") + labs(title = paste("Intercept=", signif(model_2$coefficients[[1]]),"Slope=", signif(model_2$coefficients[[2]])))
```


Showing 90% of the reads and finding the cut off
```{r warning=FALSE, message=FALSE}
ggplot(degraded_heavy, aes(x=ReadLen, y=InsertLen)) + geom_point() + labs(x="Read Len", y="Insert Len") + theme_bw() + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.90, colour="red") + geom_smooth(method = "lm") + ylim(0,100) + xlim(0,2500) + geom_hline(yintercept = 30, colour="green") + geom_vline(xintercept = 1070, colour="green")
```

3. Read len vs Soft clip len
```{r}
ggplot(subset_final, aes(x=ReadLen, y=SoftClipLen)) + geom_point() + labs(x="Read Len", y="Soft Clip Len") + theme_bw() + ylim(0,1000) + geom_smooth(method = "lm") +geom_smooth(method = "loess", colour="green")
```

Number of reads above and below the lm line
```{r warning=FALSE, message=FALSE}
model_4 <- lm(SoftClipLen ~ ReadLen, data = subset_final)
b <- resid(model_4)
sum(b<0)
sum(b>0)
```

Showing 95% of the reads and finding the cutoff
```{r warning=FALSE, message=FALSE}
ggplot(subset_final, aes(x=ReadLen, y=SoftClipLen)) + geom_point() + labs(x="Read Len", y="Soft Clip Len") + theme_bw() + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.95, colour="red") + geom_smooth(method = "lm") + ylim(0,1000) + geom_hline(yintercept = 130, colour="green") + geom_vline(xintercept = 1170, colour="green")
```

Adding correlation, slope, p-values and intercept to the plot
Plot with ggpubr library
```{r warning=FALSE, message=FALSE}
model_3 <- lm(SoftClipLen ~ ReadLen, data = degraded_heavy)

ggscatter(degraded_heavy, x="ReadLen", y="SoftClipLen", add = "reg.line", conf.int = T, cor.coef = T, cor.method = "pearson", colour="blue") + ylim(0,1000) + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.95, colour="red") +  geom_hline(yintercept = 130, colour = "green") + geom_vline(xintercept = 1170, colour="green") + labs(title = paste("Intercept=", signif(model_3$coefficients[[1]]),"Slope=", signif(model_3$coefficients[[2]])))
```

Showing 90% of the reads and finding the cutoff
```{r warning=FALSE, message=FALSE}
ggplot(degraded_heavy, aes(x=ReadLen, y=SoftClipLen)) + geom_point() + labs(x="Read Len", y="Soft Clip Len") + theme_bw() + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.90, colour="red") + geom_smooth(method = "lm") + ylim(0,1000) + geom_hline(yintercept = 120, colour="green") + geom_vline(xintercept = 1100, colour="green")
```

Showing 75% of the reads since with this the soft clip cut off reduces below 100 nt
```{r warning=FALSE, message=FALSE}
ggplot(degraded_heavy, aes(x=ReadLen, y=SoftClipLen)) + geom_point() + labs(x="Read Len", y="Soft Clip Len") + theme_bw() + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.75, colour="red") + geom_smooth(method = "lm") + ylim(0,1000) + geom_hline(yintercept = 95, colour="green") + geom_vline(xintercept = 970, colour="green")
```

## Undegraded Sample

1. Map len vs del len
Plot with both the curves- lm and loess
```{r warning=FALSE, message=FALSE}
ggplot(undegraded, aes(x=MapLen, y=DelLen)) + geom_point() + labs(x="Map Len", y="Del Len") + theme_bw() + geom_smooth(method = "lm") + geom_smooth(method = "loess", colour="green") + xlim(0,3500) + ylim(0,350)
```

Specifying the region with the cut offs
```{r warning=FALSE, message=FALSE}
ggplot(undegraded, aes(x=MapLen, y=DelLen)) + geom_point() + labs(x="Map Len", y="Del Len") + theme_bw() + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.95, colour="red") + geom_smooth(method = "lm") + xlim(0,3500) + ylim(0,350) + geom_hline(yintercept = 137, colour="green") + geom_vline(xintercept = 1600, colour="green")
```

Adding correlation, slope, p-values and intercept to the plot
Plot with ggpubr library
```{r warning=FALSE, message=FALSE}
model_1 <- lm(DelLen ~ MapLen, data = undegraded)

ggscatter(undegraded, x="MapLen", y="DelLen", add = "reg.line", conf.int = T, cor.coef = T, cor.method = "pearson", colour="blue") + ylim(0,400) + xlim(0,3000) + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.95, colour="red") +  geom_hline(yintercept = 145, colour = "green") + geom_vline(xintercept = 1600, colour="green") + labs(title = paste("Intercept=", signif(model_1$coefficients[[1]]), "Slope=", signif(model_1$coefficients[[2]])))
```

How many genes are above or below the lm line
```{r warning=FALSE, message=FALSE}
model_1 <- lm(DelLen ~ MapLen, data = undegraded)
b <- resid(model_1)
sum(b<0)  #4724
sum(b>0)  #2276
```

2. Map len vs insert len
Plot with both the curves- lm and loess
```{r warning=FALSE, message=FALSE}
ggplot(undegraded, aes(x=MapLen, y=InsertLen)) + geom_point() + labs(x="Map Len", y="Insert Len") + theme_bw() + geom_smooth(method = "lm") + geom_smooth(method = "loess", colour="green")
```

Specifying the region with the cut offs
```{r warning=FALSE, message=FALSE}
ggplot(undegraded, aes(x=MapLen, y=InsertLen)) + geom_point() + labs(x="Map Len", y="Insert Len") + theme_bw() + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.95, colour="red") + geom_smooth(method = "lm") + xlim(0,4500) + ylim(0,300) + geom_hline(yintercept = 55, colour="green") + geom_vline(xintercept = 1670, colour="green")
```

Adding correlation, slope, p-values and intercept to the plot
Plot with ggpubr library
```{r warning=FALSE, message=FALSE}
model_2 <- lm(InsertLen ~ MapLen, data = undegraded)

ggscatter(undegraded, x="MapLen", y="InsertLen", add = "reg.line", conf.int = T, cor.coef = T, cor.method = "pearson", colour="blue") + ylim(0,100) + xlim(0,2500) + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.95, colour="red") +  geom_hline(yintercept = 51, colour = "green") + geom_vline(xintercept = 1570, colour="green") + labs(title = paste("Intercept=", signif(model_2$coefficients[[1]]),"Slope=", signif(model_2$coefficients[[2]])))
```


How many genes are above or below the lm line
```{r warning=FALSE, message=FALSE}
model_1 <- lm(InsertLen ~ MapLen, data = undegraded)
b <- resid(model_1)
sum(b<0)  #4501
sum(b>0)  #2499
```

3. Map len vs soft clip len
Plot with both the curves- lm and loess
```{r warning=FALSE, message=FALSE}
ggplot(undegraded, aes(x=MapLen, y=SoftClipLen)) + geom_point() + labs(x="Map Len", y="Soft Clip Len") + theme_bw() + geom_smooth(method = "lm") + geom_smooth(method = "loess", colour="green") + xlim(0,5000) + ylim(0,1500)
```

Specifying the region with the cut offs
```{r warning=FALSE, message=FALSE}
ggplot(undegraded, aes(x=MapLen, y=SoftClipLen)) + geom_point() + labs(x="Map Len", y="Soft Clip Len") + theme_bw() + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.95, colour="red") + geom_smooth(method = "lm") + xlim(0,5000) + ylim(0,1500) + geom_hline(yintercept = 180, colour="green") + geom_vline(xintercept = 1750, colour="green")
```

How many genes are above or below the lm line
```{r warning=FALSE, message=FALSE}
model_1 <- lm(SoftClipLen ~ MapLen, data = undegraded)
b <- resid(model_1)
sum(b<0)  #5076
sum(b>0)  #1924
```

Adding correlation, slope, p-values and intercept to the plot
Plot with ggpubr library
```{r warning=FALSE, message=FALSE}
model_3 <- lm(SoftClipLen ~ MapLen, data = undegraded)

ggscatter(undegraded, x="MapLen", y="SoftClipLen", add = "reg.line", conf.int = T, cor.coef = T, cor.method = "pearson", colour="blue") + ylim(0,1000) + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.95, colour="red") +  geom_hline(yintercept = 180, colour = "green") + geom_vline(xintercept = 1750, colour="green") + labs(title = paste("Intercept=", signif(model_3$coefficients[[1]]),"Slope=", signif(model_3$coefficients[[2]])))
```



## Read len
1. Read len vs del len
```{r warning=FALSE, message=FALSE}
ggplot(undegraded, aes(x=ReadLen, y=DelLen)) + geom_point() + labs(x="Read Len", y="Del Len") + theme_bw() + geom_smooth(method = "lm") + geom_smooth(method = "loess", colour="green") +xlim(0,4000) + ylim(0,400)
```

Number of genes above or below the lm line
```{r warning=FALSE, message=FALSE}
model_3 <- lm(DelLen ~ ReadLen, data = undegraded)
b <- resid(model_3)
sum(b<0) #4836
sum(b>0) #2164
```

Showing 95% of the reads and finding the cut off
```{r warning=FALSE, message= FALSE}
ggplot(undegraded, aes(x=ReadLen, y=DelLen)) + geom_point() + labs(x="Read Len", y=" Del Len") + theme_bw() + geom_smooth(method = "lm") + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.95, colour="red") + xlim(0,4000) + ylim(0,400) + geom_hline(yintercept = 140, colour="green") + geom_vline(xintercept = 1800, colour="green")
```

Adding correlation, slope, p-values and intercept to the plot
Plot with ggpubr library
```{r warning=FALSE, message=FALSE}
model_1 <- lm(DelLen ~ ReadLen, data = undegraded)

ggscatter(undegraded, x="ReadLen", y="DelLen", add = "reg.line", conf.int = T, cor.coef = T, cor.method = "pearson", colour="blue") + ylim(0,400) + xlim(0,3500) + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.95, colour="red") +  geom_hline(yintercept = 150, colour = "green") + geom_vline(xintercept = 1800, colour="green") + labs(title = paste("Intercept=", signif(model_1$coefficients[[1]]), "Slope=", signif(model_1$coefficients[[2]])))
```


2. Read len and Insert len
```{r warning=FALSE, message=FALSE}
ggplot(undegraded, aes(x=ReadLen, y=InsertLen)) + geom_point() + labs(x="Read Len", y="Insert Len") + theme_bw()+  geom_smooth(method = "lm") + geom_smooth(method = "loess", colour="green") + xlim(0,4500) + ylim(0,300)
```

Number of genes above and below the lm line
```{r warning=FALSE, message=FALSE}
model_3 <- lm(InsertLen ~ ReadLen, data = undegraded)
b <- resid(model_3)
sum(b<0) #4307
sum(b>0) #2693
```

Showing 95% of the reads and finding the cut off
```{r warning=FALSE, message=FALSE}
ggplot(undegraded, aes(x=ReadLen, y=InsertLen)) + geom_point() + labs(x="Read Len", y="Insert Len") + theme_bw() + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.95, colour="red") + geom_smooth(method = "lm") + xlim(0,4500) +ylim(0,300) +geom_hline(yintercept = 55, colour="green") + geom_vline(xintercept = 1840, colour="green")
```

Adding correlation, slope, p-values and intercept to the plot
Plot with ggpubr library
```{r warning=FALSE, message=FALSE}
model_2 <- lm(InsertLen ~ ReadLen, data = undegraded)

ggscatter(undegraded, x="ReadLen", y="InsertLen", add = "reg.line", conf.int = T, cor.coef = T, cor.method = "pearson", colour="blue") + ylim(0,100) + xlim(0,2500) + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.95, colour="red") +  geom_hline(yintercept = 51, colour = "green") + geom_vline(xintercept = 1720, colour="green") + labs(title = paste("Intercept=", signif(model_2$coefficients[[1]]),"Slope=", signif(model_2$coefficients[[2]])))
```

3. Read len vs Soft clip len
```{r warning=FALSE, message=FALSE}
ggplot(undegraded, aes(x=ReadLen, y=SoftClipLen)) + geom_point() + labs(x="Read Len", y="Soft Clip Len") + theme_bw() + geom_smooth(method = "lm") +geom_smooth(method = "loess", colour="green") + ylim(0,1000) + xlim(0,5000)
```

Number of reads above and below the lm line
```{r warning=FALSE, message=FALSE}
model_4 <- lm(SoftClipLen ~ ReadLen, data = undegraded)
b <- resid(model_4)
sum(b<0) #4561
sum(b>0) #2439
```

Showing 95% of the reads and finding the cutoff
```{r warning=FALSE, message=FALSE}
ggplot(undegraded, aes(x=ReadLen, y=SoftClipLen)) + geom_point() + labs(x="Read Len", y="Soft Clip Len") + theme_bw() + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.95, colour="red") + geom_smooth(method = "lm") + xlim(0,5000) + ylim(0,1000) + geom_hline(yintercept = 180, colour="green") + geom_vline(xintercept = 1850, colour="green")
```

Adding correlation, slope, p-values and intercept to the plot
Plot with ggpubr library
```{r warning=FALSE, message=FALSE}
model_3 <- lm(SoftClipLen ~ ReadLen, data = undegraded)

ggscatter(undegraded, x="ReadLen", y="SoftClipLen", add = "reg.line", conf.int = T, cor.coef = T, cor.method = "pearson", colour="blue") + ylim(0,1000) + xlim(0,3500) + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.95, colour="red") +  geom_hline(yintercept = 180, colour = "green") + geom_vline(xintercept = 1850, colour="green") + labs(title = paste("Intercept=", signif(model_3$coefficients[[1]]),"Slope=", signif(model_3$coefficients[[2]])))
```


## Mildly Degraded Sample

I. MAP LEN

Adding correlation, slope, p-values and intercept to the plot
Plot with ggpubr library
```{r warning=FALSE, message=FALSE}
model_1 <- lm(DelLen ~ MapLen, data = mildly_degraded)

ggscatter(mildly_degraded, x="MapLen", y="DelLen", add = "reg.line", conf.int = T, cor.coef = T, cor.method = "pearson", colour="blue") + ylim(0,400) + xlim(0,3000) + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.95, colour="red") +  geom_hline(yintercept = 70, colour = "green") + geom_vline(xintercept = 1100, colour="green") + labs(title = paste("Intercept=", signif(model_1$coefficients[[1]]), "Slope=", signif(model_1$coefficients[[2]])))
```

Adding correlation, slope, p-values and intercept to the plot
Plot with ggpubr library
```{r warning=FALSE, message=FALSE}
model_2 <- lm(InsertLen ~ MapLen, data = mildly_degraded)

ggscatter(mildly_degraded, x="MapLen", y="InsertLen", add = "reg.line", conf.int = T, cor.coef = T, cor.method = "pearson", colour="blue") + ylim(0,100) + xlim(0,2500) + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.95, colour="red") +  geom_hline(yintercept = 37, colour = "green") + geom_vline(xintercept = 1080, colour="green") + labs(title = paste("Intercept=", signif(model_2$coefficients[[1]]),"Slope=", signif(model_2$coefficients[[2]])))
```


Adding correlation, slope, p-values and intercept to the plot
Plot with ggpubr library
```{r warning=FALSE, message=FALSE}
model_3 <- lm(SoftClipLen ~ MapLen, data = mildly_degraded)

ggscatter(mildly_degraded, x="MapLen", y="SoftClipLen", add = "reg.line", conf.int = T, cor.coef = T, cor.method = "pearson", colour="blue") + ylim(0,1000) + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.95, colour="red") +  geom_hline(yintercept = 135, colour = "green") + geom_vline(xintercept = 1130, colour="green") + labs(title = paste("Intercept=", signif(model_3$coefficients[[1]]),"Slope=", signif(model_3$coefficients[[2]])))
```


II. Read Len

Adding correlation, slope, p-values and intercept to the plot
Plot with ggpubr library
```{r warning=FALSE, message=FALSE}
model_1 <- lm(DelLen ~ ReadLen, data = mildly_degraded)

ggscatter(mildly_degraded, x="ReadLen", y="DelLen", add = "reg.line", conf.int = T, cor.coef = T, cor.method = "pearson", colour="blue") + ylim(0,400) + xlim(0,3500) + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.95, colour="red") +  geom_hline(yintercept = 70, colour = "green") + geom_vline(xintercept = 1200, colour="green") + labs(title = paste("Intercept=", signif(model_1$coefficients[[1]]), "Slope=", signif(model_1$coefficients[[2]])))
```

Adding correlation, slope, p-values and intercept to the plot
Plot with ggpubr library
```{r warning=FALSE, message=FALSE}
model_2 <- lm(InsertLen ~ ReadLen, data = mildly_degraded)

ggscatter(mildly_degraded, x="ReadLen", y="InsertLen", add = "reg.line", conf.int = T, cor.coef = T, cor.method = "pearson", colour="blue") + ylim(0,100) + xlim(0,2500) + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.95, colour="red") +  geom_hline(yintercept = 37, colour = "green") + geom_vline(xintercept = 1180, colour="green") + labs(title = paste("Intercept=", signif(model_2$coefficients[[1]]),"Slope=", signif(model_2$coefficients[[2]])))
```

Adding correlation, slope, p-values and intercept to the plot
Plot with ggpubr library
```{r warning=FALSE, message=FALSE}
model_3 <- lm(SoftClipLen ~ ReadLen, data = mildly_degraded)

ggscatter(mildly_degraded, x="ReadLen", y="SoftClipLen", add = "reg.line", conf.int = T, cor.coef = T, cor.method = "pearson", colour="blue") + ylim(0,1000) + xlim(0,3500) + stat_ellipse(geom = "polygon", alpha=0.2, level = 0.95, colour="red") +  geom_hline(yintercept = 140, colour = "green") + geom_vline(xintercept = 1200, colour="green") + labs(title = paste("Intercept=", signif(model_3$coefficients[[1]]),"Slope=", signif(model_3$coefficients[[2]])))
```


Plots of max map, del, insert and soft clip len
MaxDel len
```{r warning=FALSE, message=FALSE}
merge_cigar$MaxDelLen[!is.finite(merge_cigar$MaxDelLen)] <- 0

model_1 <- lm(MaxDelLen ~ MapLen, data = merge_cigar)

ggplot(merge_cigar, aes(x=MapLen, y=MaxDelLen)) + geom_point(pch=1) + stat_cor(method = "pearson") + geom_smooth(method = "lm", col="blue") + ggtitle("Heavily degraded") + theme_bw() + ylim(0,750)

ggplot(merge_cigar, aes(x=MapLen, y=MaxDelLen)) + geom_point(pch=1) + stat_cor(method = "pearson") + geom_smooth(method = "lm", col="blue") + ggtitle("Mildly degraded") + theme_bw() + ylim(0,750)

ggplot(merge_cigar, aes(x=MapLen, y=MaxDelLen)) + geom_point(pch=1) + stat_cor(method = "pearson") + geom_smooth(method = "lm", col="blue") + ggtitle("Undegraded") + theme_bw() + ylim(0,750)
```

MaxInsert Len
```{r warning=FALSE, message=FALSE}
merge_cigar$MaxInsertLen[!is.finite(merge_cigar$MaxInsertLen)] <- 0

model_2 <- lm(MaxInsertLen ~ MapLen, data = merge_cigar)

ggplot(merge_cigar, aes(x=MapLen, y=MaxInsertLen)) + geom_point(pch=1) + stat_cor(method = "pearson") + geom_smooth(method = "lm", col="blue") + ggtitle("Heavily degraded") + theme_bw() + ylim(0,500)

ggplot(merge_cigar, aes(x=MapLen, y=MaxInsertLen)) + geom_point(pch=1) + stat_cor(method = "pearson") + geom_smooth(method = "lm", col="blue") + ggtitle("Mildly degraded") + theme_bw() + ylim(0,500)

ggplot(merge_cigar, aes(x=MapLen, y=MaxInsertLen)) + geom_point(pch=1) + stat_cor(method = "pearson") + geom_smooth(method = "lm", col="blue") + ggtitle("Undegraded") + theme_bw() + ylim(0,500)
```

MaxSoftClipLen
```{r warning=FALSE, message=FALSE}
merge_cigar$MaxSoftClipLen[!is.finite(merge_cigar$MaxSoftClipLen)] <- 0

model_3 <- lm(MaxSoftClipLen ~ MapLen, data = merge_cigar)

ggplot(merge_cigar, aes(x=MapLen, y=MaxSoftClipLen)) + geom_point(pch=1) + stat_cor(method = "pearson") + geom_smooth(method = "lm", col="blue") + ggtitle("Heavily degraded") + theme_bw()

ggplot(merge_cigar, aes(x=MapLen, y=MaxSoftClipLen)) + geom_point(pch=1) + stat_cor(method = "pearson") + geom_smooth(method = "lm", col="blue") + ggtitle("Mildly degraded") + theme_bw() + xlim(0,4500) + ylim(0,4500)

ggplot(merge_cigar, aes(x=MapLen, y=MaxSoftClipLen)) + geom_point(pch=1) + stat_cor(method = "pearson") + geom_smooth(method = "lm", col="blue") + ggtitle("Undegraded") + theme_bw() + xlim(0,4500) + ylim(0,4500)
```

